src/gwt/src/org/rstudio/studio/client/workbench/views/terminal/TerminalPane.java

```diff
@@ -0,0 +1,256 @@
+import org.rstudio.studio.client.common.console.ConsoleProcess;
+import org.rstudio.studio.client.common.console.ConsoleProcessInfo;
+import org.rstudio.studio.client.common.console.ProcessExitEvent;
+import org.rstudio.studio.client.common.crypto.CryptoServerOperations;
+import org.rstudio.studio.client.common.crypto.PublicKeyInfo;
+import org.rstudio.studio.client.common.crypto.RSAEncrypt;
+import org.rstudio.studio.client.common.shell.ShellInput;
+import org.rstudio.studio.client.server.ServerError;
+import org.rstudio.studio.client.server.ServerRequestCallback;
+import org.rstudio.studio.client.server.Void;
+import org.rstudio.studio.client.server.VoidServerRequestCallback;
+import org.rstudio.studio.client.workbench.model.WorkbenchServerOperations;
+import org.rstudio.studio.client.workbench.ui.WorkbenchPane;
+import org.rstudio.studio.client.workbench.views.terminal.events.ResizeTerminalEvent;
+import org.rstudio.studio.client.workbench.views.terminal.events.TerminalDataInputEvent;
+
+import com.google.gwt.event.dom.client.ClickEvent;
+import com.google.gwt.event.dom.client.ClickHandler;
+import com.google.gwt.event.shared.HandlerRegistration;
+import com.google.gwt.user.client.Command;
+import com.google.gwt.user.client.ui.ResizeLayoutPanel;
+import com.google.gwt.user.client.ui.Widget;
+
+public class TerminalPane extends WorkbenchPane
+                          implements ConsoleOutputEvent.Handler, 
+                                     ProcessExitEvent.Handler,
+                                     ClickHandler,
+                                     ResizeTerminalEvent.Handler,
+                                     TerminalDataInputEvent.Handler
+{
+   protected TerminalPane(String title, WorkbenchServerOperations server)
+   {
+      super(title);
+      server_ = server;
+      host_ = new ResizeLayoutPanel();
+   }
+
+   @Override
+   protected Widget createMainWidget()
+   {
+      XTermWidget.load(new Command()
+      {
+         @Override
+         public void execute()
+         {
+            xterm_ = new XTermWidget();
+            xterm_.setHeight("100%");
+            host_.add(xterm_);
+            connectToTerminalProcess();
+         }
+      });
+      return host_;
+   }
+  
+   /**
+    * Create a terminal process and connect to it.
+    */
+   private void connectToTerminalProcess()
+   {
+      server_.startShellDialog(ConsoleProcess.TerminalType.XTERM, 
+                               80, 25,
+                               new ServerRequestCallback<ConsoleProcess>()
+      {
+         @Override
+         public void onResponseReceived(ConsoleProcess consoleProcess)
+         {
+            consoleProcess_ = consoleProcess;
+            cryptoServer_ = server_;
+            
+            if (getInteractionMode() != ConsoleProcessInfo.INTERACTION_ALWAYS)
+            {
+               throw new IllegalArgumentException("Unsupport ConsoleProcess interaction mode");
```

Comment:
Where is this likely to be caught? If we fail to launch the terminal for some reason we still need to respond to the user's gesture in some way (even if it's just to show a dialog or error bar that says "sorry, this didn't work").